lib/rules/media-feature-name-value-no-unknown/index.js

```diff
@@ -0,0 +1,308 @@
+
+				if (supportedValueTypes.has('length') && lengthUnits.has(unit)) return;
+
+				// An unexpected dimension or a media feature that doesn't support dimensions.
+				reporter(state, raw, start, end);
+			}
+		}
+
+		/**
+		 * Check that a function node is valid for a given media feature name.
+		 *
+		 * @param {State} state
+		 * @param {import('@csstools/css-parser-algorithms').FunctionNode} functionNode
+		 * @param {Reporter} reporter
+		 * @returns {void}
+		 */
+		function checkFunction(state, functionNode, reporter) {
+			const functionName = functionNode.getName().toLowerCase();
+
+			// "env()" can represent any value, it is treated as valid for static analysis.
+			if (functionName === 'env') return;
+
+			const supportedValueTypes = mediaFeatureNameAllowedValueTypes.get(state.mediaFeatureName);
+
+			if (
+				supportedValueTypes &&
+				mathFunctions.has(functionName) &&
+				(supportedValueTypes.has('integer') ||
+					supportedValueTypes.has('length') ||
+					supportedValueTypes.has('ratio') ||
+					supportedValueTypes.has('resolution'))
+			) {
+				return;
+			}
+
+			// An unexpected function or a media feature that doesn't support types that can be the result of a function.
+			reporter(state, functionNode.toString(), ...startAndEndIndex(functionNode));
+		}
+
+		/**
+		 * Check that an array of component values is valid for a given media feature name.
+		 *
+		 * @param {State} state
+		 * @param {Array<import('@csstools/css-parser-algorithms').ComponentValue>} componentValues
+		 * @param {Reporter} reporter
+		 * @returns {void}
+		 */
+		function checkListOfComponentValues(state, componentValues, reporter) {
+			const supportedValueTypes = mediaFeatureNameAllowedValueTypes.get(state.mediaFeatureName);
+
+			if (
+				supportedValueTypes &&
+				supportedValueTypes.has('ratio') &&
+				matchesRatioExactly(componentValues) !== -1
+			) {
+				return;
+			}
+
+			// An invalid aspect ratio or a media feature that doesn't support aspect ratios.
+			reporter(
+				state,
+				componentValues.map((x) => x.toString()).join(''),
+				...startAndEndIndex(componentValues),
+			);
+		}
+
+		/**
+		 * @param {State} state
+		 * @param {import('@csstools/media-query-list-parser').MediaFeatureValue} valueNode
+		 * @param {Reporter} reporter
+		 * @returns {void}
+		 */
+		function checkMediaFeatureValue(state, valueNode, reporter) {
+			if (isTokenNode(valueNode.value)) {
+				checkSingleToken(state, valueNode.value.value, reporter);
+
+				return;
+			}
+
+			if (isFunctionNode(valueNode.value)) {
+				checkFunction(state, valueNode.value, reporter);
+
+				return;
+			}
+
+			if (Array.isArray(valueNode.value)) {
+				checkListOfComponentValues(state, valueNode.value, reporter);
+			}
+		}
+
+		root.walkAtRules(/^media$/i, (atRule) => {
+			/**
+			 * @type {Reporter}
+			 */
+			const reporter = (state, valuePart, start, end) => {
+				const atRuleParamIndexValue = atRuleParamIndex(atRule);
+
```

Comment:
[suggest] Does it make sense to store `atRuleParamIndexValue` like `atRule` in a closure to reduce calculation?

```suggestion
			const atRuleParamIndexValue = atRuleParamIndex(atRule);

			/** @type {Reporter} */
			const reporter = (state, valuePart, start, end) => {
```

Reply:
I went back on forth on this.
I think your proposed change would reduce calculations when the number of warnings is more than 1, right?

The current code reduces calculations when there are no warnings at all.
I think this will lead to fewer calculations overall, but happy to make the change :)

Reply:
> The current code reduces calculations when there are no warnings at all.

Okay, leaving it as-is makes sense. Thanks for the explanation! 👍🏼